mmgen/models/architectures/sngan_proj/modules.py

```diff
@@ -0,0 +1,236 @@
+from copy import deepcopy
+
+import numpy as np
+import torch.nn as nn
+from mmcv.cnn import (ConvModule, build_activation_layer, build_norm_layer,
+                      build_upsample_layer, constant_init, xavier_init)
+
+from mmgen.models.builder import MODULES
+from mmgen.utils import check_dist_init
+
+
+@MODULES.register_module()
+class SNGAN_GenResBlock(nn.Module):
+
+    # _default_conv_cfg = dict(kernel=3, stride=1, padding=1, act_cfg=None)
+    # _default_shortcut_cfg = dict(kernel=1, stride=1, padding=0, act_cfg=None)
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 hidden_channels=None,
+                 num_classes=0,
+                 use_cbn=True,
+                 use_norm_para=False,
+                 act_cfg=dict(type='ReLU'),
+                 norm_cfg=dict(type='BN'),
+                 upsample_cfg=dict(type='nearest'),
+                 upsample=True,
+                 auto_sync_bn=True,
+                 conv_cfg=None,
+                 shortcut_cfg=None):
+
+        super().__init__()
+        self.learnable_sc = in_channels != out_channels or upsample
+
+        self.activateion = build_activation_layer(act_cfg)
+        hidden_channels = out_channels if hidden_channels is None \
+            else hidden_channels
+
+        if upsample:
+            self.upsample = build_upsample_layer(upsample_cfg)
+
+        conv_blocks = [
+            ConvModule(in_channels, hidden_channels, **conv_cfg),
```

Comment:
Please pay attention to the default behavior of `ConvModule`. Thus, we may also provide the default conv config for `conv_cfg`. In addition, I may recommend directly offer the API of `with_spectral_norm` like:

```python
class ResBlock(self, xx, xx, with_spectral_norm=False):
    ConvModule(in_channels, hidden_channels, with_spectral_norm=with_spectral_norm, **conv_cfg)
```